odl/deform/linear_deformation.py

```diff
@@ -0,0 +1,377 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Shape-based reconstruction using optimal information transportation.
+
+The Fisher-Rao metric is used in regularization term. And L2 data matching
+term is used in fitting term.
+"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+from odl.operator import Operator
+from odl.discr import DiscreteLp, Divergence
+import odl
+
+__all__ = ('LinearDeformation', 'MassPreservingLinearDeformation',
+           'LinearizedDeformationOperator',
+           'MassPreservingDeformationOperator')
+
+
+class LinearizedDeformationOperator(Operator):
+    """
+    Linearized deformation operator mapping parameters into a
+    deformed template. It is a left action given by pushforward.
+
+    This operator computes the deformed template for a fixed
+    template ``I``:
+
+        v --> I(Id + v)
+
+    where ``Id`` is the identity mapping:
+
+        y --> y
+
+    and the vector field ``v`` is a displacement mapping:
+
+        y --> v(y)
+
+    Here, the ``y`` is an element in the domain of the target.
+    In diffeomorphic sense,  the domain of ``I`` is the same
+    as the target.
+    """
+
+    def __init__(self, template):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector`
+            Fixed template deformed by the vector field.
+        """
+        self.template = template
+        self.domain_space = odl.ProductSpace(self.template.space,
+                                             self.template.space.ndim)
+
+        super().__init__(self.domain_space, self.template.space, linear=False)
+
+    def _call(self, displacement):
+        """Implementation of ``self(displacement)``.
+
+        Parameters
+        ----------
+        displacement : `ProductSpaceVector`
+            Linearized deformation parameters for image grid points.
+        """
+        image_pts = self.template.space.grid.points()
+        image_pts += np.asarray(displacement).T
```

Comment:
This is extremely slow since it will access the displacement vector element by element.